Gulpfile.ts

```diff
@@ -0,0 +1,1125 @@
+/// <reference path="scripts/types/ambient.d.ts" />
+
+import * as cp from "child_process";
+import * as path from "path";
+import * as fs from "fs";
+import originalGulp = require("gulp");
+import helpMaker = require("gulp-help");
+import runSequence = require("run-sequence");
+import concat = require("gulp-concat");
+import clone = require("gulp-clone");
+import tsc = require("gulp-typescript");
+declare module "gulp-typescript" {
+    interface Settings {
+        stripInternal?: boolean;
+        newLine?: number;
+    }
+    interface CompileStream extends NodeJS.ReadWriteStream {} // Either gulp or gulp-typescript has some odd typings which don't reflect reality, making this required
+}
+import * as insert from "gulp-insert";
+import * as sourcemaps from "gulp-sourcemaps";
+declare global {
+    // This is silly. We include Q because orchestrator (a part of gulp) depends on it, but its not included.
+    // `del` further depends on `Promise` (and is also not included), so we just, patch the global scope's Promise to Q's
+    type Promise<T> = Q.Promise<T>;
+}
+import del = require("del");
+import mkdirP = require("mkdirP");
+import minimist = require("minimist");
+import browserify = require("browserify");
+import through2 = require("through2");
+import intoStream = require("into-stream");
+import * as os from "os";
+import Linter = require("tslint");
+const gulp = helpMaker(originalGulp);
+const mochaParallel = require("./scripts/mocha-parallel.js");
+const {runTestsInParallel} = mochaParallel;
+
+const cmdLineOptions = minimist(process.argv.slice(2), {
+    boolean: ["debug", "light", "colors", "lint", "soft"],
+    string: ["browser", "tests", "host", "reporter", ],
+    alias: {
+        d: "debug",
+        t: "tests",
+        test: "tests",
+        r: "reporter",
+        color: "colors",
+        f: "files",
+        file: "files"
+    },
+    default: {
+        soft: false,
+        colors: process.env.colors || process.env.color || true,
+        debug: process.env.debug || process.env.d,
+        host: process.env.TYPESCRIPT_HOST || process.env.host || "node",
+        browser: process.env.browser || process.env.b || "IE",
+        tests: process.env.test || process.env.tests || process.env.t,
+        light: process.env.light || false,
+        port: process.env.port || process.env.p || "8888",
+        reporter: process.env.reporter || process.env.r,
+        lint: process.env.lint || true,
+        files: process.env.f || process.env.file || process.env.files || "",
+    }
+});
+
+function exec(cmd: string, args: string[], complete: () => void = (() => {}), error: (e: any, status: number) => void = (() => {})) {
+    console.log(`${cmd} ${args.join(" ")}`);
+    const ex = cp.spawn(cmd, args);

```

Comment:
Also if you do:
```ts
const ex = cp.spawn(win ? "cmd" : "/bin/sh", [win ? "/c" : "-c", cmd, ...args], { stdio: "inherit", windowsVerbatimArguments: true });
```
It will use the OS shell to launch the command, letting `spawn` leverage the `PATH` environment variable and automatically handle executables without appending `.cmd`, etc.
```